File: 0207-course-schedule/0207-course-schedule.cpp
class Solution {
public:
    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
        vector<vector<int>> graph(numCourses);
        vector<int> inDegree(numCourses,0);
        for(auto edge:prerequisites){
            graph[edge[1]].push_back(edge[0]);
            inDegree[edge[0]]++;
        }
        int count = 0;
        queue<int> q;
        for(int i=0;i<numCourses;i++){
            if(inDegree[i]==0)
                q.push(i);
        }
        while(!q.empty()){
            auto temp = q.front();
            q.pop();
            ++count;
            for(auto v: graph[temp]){
                if(--inDegree[v]==0){
                    q.push(v);
                }
            }
        }
        return count == numCourses;
    }
};

File: 0207-course-schedule/0207-course-schedule.py
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        if numCourses <= 0:
            return False
        inDegree = {i : 0 for i in range(numCourses)}
        graph = {i : [] for i in range(numCourses)}
        
        for child, parent in prerequisites:
            graph[parent].append(child)
            inDegree[child] += 1

        sources = deque()
        
        for key in inDegree:
            if inDegree[key] == 0:
                sources.append(key)
        visited = 0       
        while sources:
            vertex = sources.popleft()
            visited += 1
            for child in graph[vertex]:
                inDegree[child] -= 1
                if inDegree[child] == 0:
                    sources.append(child)
        
        return visited == numCourses